File: scripts/test_solutions/417_pacific_atlantic_water_flow.js
var pacificAtlantic = function(heights) {
    const m = heights.length, n = heights[0].length;
    const pacific = Array(m).fill(null).map(() => Array(n).fill(false));
    const atlantic = Array(m).fill(null).map(() => Array(n).fill(false));
    function dfs(i, j, ocean) {
        ocean[i][j] = true;
        const dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]];
        for (const [di, dj] of dirs) {
            const ni = i + di, nj = j + dj;
            if (ni >= 0 && ni < m && nj >= 0 && nj < n && !ocean[ni][nj] && heights[ni][nj] >= heights[i][j]) dfs(ni, nj, ocean);
        }
    }
    for (let i = 0; i < m; i++) { dfs(i, 0, pacific); dfs(i, n - 1, atlantic); }
    for (let j = 0; j < n; j++) { dfs(0, j, pacific); dfs(m - 1, j, atlantic); }
    const result = [];
    for (let i = 0; i < m; i++) for (let j = 0; j < n; j++) if (pacific[i][j] && atlantic[i][j]) result.push([i, j]);
    return result;
};

File: scripts/test_solutions/417_pacific_atlantic_water_flow.cpp
class Solution {
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        if (heights.empty()) return {};
        int m = heights.size(), n = heights[0].size();

        vector<vector<bool>> pacific(m, vector<bool>(n, false));
        vector<vector<bool>> atlantic(m, vector<bool>(n, false));

        for (int i = 0; i < m; i++) {
            dfs(heights, pacific, i, 0);
            dfs(heights, atlantic, i, n - 1);
        }
        for (int j = 0; j < n; j++) {
            dfs(heights, pacific, 0, j);
            dfs(heights, atlantic, m - 1, j);
        }

        vector<vector<int>> result;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (pacific[i][j] && atlantic[i][j]) {
                    result.push_back({i, j});
                }
            }
        }
        return result;
    }

private:
    int dirs[4][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    void dfs(vector<vector<int>>& heights, vector<vector<bool>>& ocean, int i, int j) {
        int m = heights.size(), n = heights[0].size();
        ocean[i][j] = true;
        for (auto& d : dirs) {
            int ni = i + d[0], nj = j + d[1];
            if (ni >= 0 && ni < m && nj >= 0 && nj < n && !ocean[ni][nj] && heights[ni][nj] >= heights[i][j]) {
                dfs(heights, ocean, ni, nj);
            }
        }
    }
};
